scripts/ci/notification/matrix_notification.sh

```diff
@@ -0,0 +1,107 @@
+        --header "Authorization: Bearer ${MATRIX_ACCESS_TOKEN}" \
+        --data "$formatted_message"
+}
+
+circleci_workflow_name() {
+    # Workflow name is not exposed as an env variable. Has to be queried from the API.
+    # The name is not critical so if anything fails, use the raw workflow ID as a fallback.
+    workflow_info=$(curl --silent "https://circleci.com/api/v2/workflow/${CIRCLE_WORKFLOW_ID}") || true
+    workflow_name=$(echo "$workflow_info" | grep -E '"\s*name"\s*:\s*".*"' | cut -d \" -f 4 || echo "$CIRCLE_WORKFLOW_ID")
+    printf '%s' "$(sanitize_variable "$workflow_name")"
+}
+
+circleci_job_name() {
+    [[ $CIRCLE_NODE_TOTAL == 1 ]] && job="${CIRCLE_JOB}"
+    [[ $CIRCLE_NODE_TOTAL != 1 ]] && job="${CIRCLE_JOB} (run $((CIRCLE_NODE_INDEX + 1))/${CIRCLE_NODE_TOTAL})"
+    printf '%s' "$(sanitize_variable "$job")"
+}
+
+# Currently matrix only supports html formatted body,
+# see: https://spec.matrix.org/v1.6/client-server-api/#mtext
+#
+# Eventually, the matrix api will have support for better format options and `formatted_body` may not be necessary anymore:
+# https://github.com/matrix-org/matrix-spec-proposals/pull/1767
+format_message() {
+    [[ -z "$1" ]] && fail "Event type not provided."
+    local event="$1"
+
+    [[ "$event" == "failure" ]] && message="$(cat "${SCRIPT_DIR}/templates/build_fail.json")"
+    [[ "$event" == "success" ]] && message="$(cat "${SCRIPT_DIR}/templates/build_success.json")"
+    [[ "$event" == "release" ]] && message="$(cat "${SCRIPT_DIR}/templates/build_release.json")"
+
+    [[ -z "$message" ]] && fail "Message for event [$event] not defined."
+
+    # Escape \ and "
+    formatted_msg="$(printf '%s' "$message" | sed 's/\\/\\\\/g' | sed 's/"/\\"/g')"
+    # Evaluate environment variables
+    # NOTE: The variables must be sanitized before call eval to avoid execution
+    # of unexpected commands instead of valid input values.
+    # shellcheck disable=SC2005
+    echo "$(eval printf '%s' \""${formatted_msg}"\")"
+}
+
+sanitize_variable() {
+    # Filter special shell symbols like: $, {, }, `(backtick), \(backslash), etc
+    # to prevent command substitution.
+    #
+    # The regex below allows ONLY:
+    # :blank: - space and tab
+    # :alnum: - the same as: [0-9A-Za-z]
+    # And the symbols (excluding comma):
+    #   _, :, / (forward slash), ?, =, &, . (dot), (, ), -
+    #
+    # Some of those characters were allowed to correctly preserve URLs.
+    # All other characters that are not in this list are removed.
+    #
+    # See: https://www.gnu.org/software/sed/manual/html_node/Character-Classes-and-Bracket-Expressions.html
+    sanitized_value="$(printf '%s' "$1" | sed "s;[^[:blank:][:alnum:]_:/?=&.()-];;g")"
+    echo "$sanitized_value"
+}
+
+# Set message environment variables based on CI backend
+if [[ "$CIRCLECI" = true ]] ; then
+    BRANCH="$CIRCLE_BRANCH"
+    TAG="$CIRCLE_TAG"
+    BUILD_URL="$CIRCLE_BUILD_URL"
+    BUILD_NUM="$CIRCLE_BUILD_NUM"
+fi
```

Comment:
Not sure there's any point in doing this. You're still using other CircleCI variables like `$CIRCLE_NODE_INDEX`, comments and function names refer to CircleCI and overall we're pretty much assuming we're working with CircleCI. Might make more sense to just explicitly fail when `$CIRCLECI` is not `"true"` and keep the script simpler, with the assumption will actually make it work with whatever other CI we need when we need it.

Also, if you really want to keep it CI-independent, I think that passing these values to the function via arguments would be nicer.

Reply:
After your tweaks this now seems less to be dependent only on these variables (other than maybe the fact that it always runs CircleCI-specific workarounds).

I still think it would be better to do this only when we need this, but if you prefer it that way it's fine too.

Reply:
Yeah, but the fix of #13507 is trivial with the changes of this PR here, basically replacing the environment variables, or I'm missing something there haha

By CircleCI-specific workarounds you mean this comment https://github.com/ethereum/solidity/pull/14090#discussion_r1156943573, right? Actually, I still need to check what was this about, I don't really got what is the problem from the FIXME comment.

I mean, if CircleCI associates random PRs with what was supposed to be the `CIRCLE_PULL_REQUEST` it is a bug that we should report to upstream, no?

Reply:
> Yeah, but the fix of https://github.com/ethereum/solidity/issues/13507 is trivial with the changes of this PR here

Ah, I missed that bit from your original comment. So it's not just future-proofing but you'll actually be using that extra code for portability. Fine then.

> By CircleCI-specific workarounds you mean this comment https://github.com/ethereum/solidity/pull/14090#discussion_r1156943573, right? Actually, I still need to check what was this about, I don't really got what is the problem from the FIXME comment.

See #12269. Originally I was checking if `$CIRCLE_PULL_REQUEST` is empty/unset to determine if the workflow is one running in a PR. It quickly turned out that this variable often points at a PR even in runs on `develop`.

I just checked the "Preparing environment variables" step in a few random workflows and I'm not seeing that any more. Must have been fixed by CircleCI.

> I mean, if CircleCI associates random PRs with what was supposed to be the CIRCLE_PULL_REQUEST it is a bug that we should report to upstream, no?

Yes, I did report a similar issue to their support back then: https://github.com/ethereum/solidity/pull/12202#issuecomment-952950518. I remember that I even got a reply later that this is super low priority for them and we have to live with it for the time being. And later again that they'll be fixing it after all.

Reply:
Ah nice, thanks for the reference. I will remove the `FIXME` when changing this script to fix https://github.com/ethereum/solidity/issues/13507 then. ;)